validator/bundle.py

```diff
@@ -41,6 +45,109 @@ def get_sub_type(self, discriminator: str) -> Optional["GraphqlType"]:
         return None
 
 
+class InterfaceResolve(TypedDict):
+    strategy: str
+    fieldMap: NotRequired[dict[str, str]]
+    field: NotRequired[str]
+
+
+class GraphqlField(TypedDict):
+    name: str
+    type: str
+    isUnique: NotRequired[bool]
+    isContextUnique: NotRequired[bool]
+    datafileSchema: NotRequired[str]
+
+
+@dataclass(frozen=True)
+class GraphqlTypeV2:
+    name: str
+    fields: dict[str, GraphqlField]
+    datafile: str | None
+    is_interface: bool
+    interface_resolve: InterfaceResolve | None
+
+    def get_field(self, field_name: str) -> GraphqlField | None:
+        return self.fields.get(field_name)
+
+    def is_interface_field_map(self) -> bool:
+        return (
+            self.is_interface
+            and self.interface_resolve is not None
+            and self.interface_resolve.get("strategy") == "fieldMap"
+        )
+
+    def interface_resolve_field_name(self) -> str | None:
+        if self.is_interface_field_map() and self.interface_resolve:
+            return self.interface_resolve.get("field")
+        return None
+
+    def resolve_interface_field_value(self, data: Any) -> Any:
+        if (field_name := self.interface_resolve_field_name()) and isinstance(
+            data, dict
+        ):
+            return data.get(field_name)
+        return None
+
+    def resolve_interface_type_name(self, data: Any) -> str | None:
+        if (
+            self.interface_resolve
+            and (field_map := self.interface_resolve.get("fieldMap"))
+            and (field_value := self.resolve_interface_field_value(data))
+        ):
+            return field_map.get(field_value)
+        return None
+
+
+class GraphqlLookup:
+    def __init__(self, confs: list[dict[str, Any]]):
+        self.graphql_types = {
+            name: self._build_graphql_type(conf)
+            for conf in confs
+            if (name := conf.get("name"))
+        }
+        type_name_by_type_datafile = {
+            datafile: type_name
+            for type_name, graphql_type in self.graphql_types.items()
+            if (datafile := graphql_type.datafile)
+        }
+        type_name_by_query_datafile_schema = (
+            {
+                datafile_schema: field["type"]
+                for field in query.fields.values()
+                if (datafile_schema := field.get("datafileSchema"))
+            }
+            if (query := self.graphql_types.get("Query"))
+            else {}
+        )
```

Comment:
IMHO it is difficult to understand maybe don't use all the shortcuts to do this

Reply:
extract func in https://github.com/app-sre/qontract-validator/pull/99/commits/02a274645d3b9eaa3541d43748f2e4e1877bf86a